### PR TITLE
Keep track of weight in MatchGroup

### PR DESCRIPTION
**Pull request by mpickering**
When desugaring case expressions we sometimes invent new variables which
need to be scaled like source level case expressions. Storing this
weight allows us to perform this scaling appropiately.

This fixes test `T11339c` at least.

### CODE REVIEW

File: compiler/deSugar/DsArrows.hs [hunk 1]
```diff
@@ -604,7 +604,8 @@ dsCmd ids local_vars stack_ty res_ty
     core_body <- dsExpr (HsCase noExt exp
                          (MG { mg_alts = L l matches'
                              , mg_arg_tys = arg_tys
-                             , mg_res_ty = sum_ty, mg_origin = origin }))
+                             , mg_res_ty = sum_ty, mg_origin = origin
+                             , mg_weight = Omega })) -- TODO: MattP
```
Comment (aspiwack) [inline compiler/deSugar/DsArrows.hs]: Even though a moment inspection will make it clear that the problem is a rogue `Omega`, my advice is to make a note of this. This will help when you come back here in a week or two.

File: compiler/deSugar/DsExpr.hs [hunk 2]
```diff
@@ -215,8 +215,8 @@ dsUnliftedBind (PatBind {pat_lhs = pat, pat_rhs = grhss
        ; let upat = unLoc pat
              eqn = EqnInfo { eqn_pats = [upat],
                              eqn_rhs = cantFailMatchResult body }
-       ; var    <- selectMatchVar upat
-       ; result <- matchEquations PatBindRhs [var] [eqn] (exprType body)
+       ; var    <- selectMatchVar Omega upat -- TODO: MattP
+       ; result <- matchEquations PatBindRhs [var] [eqn] (exprType body) Omega -- MattP: TODO
```
Comment (aspiwack) [inline compiler/deSugar/DsExpr.hs]: Same as previous comment.

File: compiler/deSugar/DsExpr.hs [hunk 3]
```diff
@@ -632,7 +632,9 @@ ds_expr _ expr@(RecordUpd { rupd_expr = record_expr, rupd_flds = fields
         ; ([discrim_var], matching_code)
                 <- matchWrapper RecUpd Nothing (MG { mg_alts = noLoc alts
                                                    , mg_arg_tys = [unrestricted in_ty] -- TODO: arnaud: haven't thought of the rules for record update
-                                                   , mg_res_ty = out_ty, mg_origin = FromSource })
+                                                   , mg_res_ty = out_ty
+                                                   , mg_origin = FromSource
+                                                   , mg_weight = Omega }) -- MatP: TODO
```
Comment (aspiwack) [inline compiler/deSugar/DsExpr.hs]: Same again.

File: compiler/deSugar/DsExpr.hs [hunk 4]
```diff
@@ -956,7 +958,8 @@ dsDo stmts
                                                        body']
                       , mg_arg_tys = map unrestricted arg_tys
                       , mg_res_ty = body_ty
-                      , mg_origin = Generated }
+                      , mg_origin = Generated
+                      , mg_weight = Omega }
```
Comment (aspiwack) [inline compiler/deSugar/DsExpr.hs]: Should there be a `TODO` note here too?

File: compiler/deSugar/DsExpr.hs [hunk 5]
```diff
@@ -988,7 +991,8 @@ dsDo stmts
                                                     LambdaExpr
                                                     [mfix_pat] body]
                                , mg_arg_tys = [unrestricted tup_ty], mg_res_ty = body_ty
-                               , mg_origin = Generated })
+                               , mg_origin = Generated
+                               , mg_weight = Omega })
```
Comment (aspiwack) [inline compiler/deSugar/DsExpr.hs]: What about here?

File: compiler/deSugar/DsUtils.hs [hunk 6]
```diff
@@ -98,7 +98,7 @@ otherwise, make one up.
 -}
 
 selectSimpleMatchVarL :: LPat GhcTc -> DsM Id
-selectSimpleMatchVarL pat = selectMatchVar (unLoc pat)
+selectSimpleMatchVarL pat = selectMatchVar Omega (unLoc pat) -- TODO: MattP
```
Comment (aspiwack) [inline compiler/deSugar/DsUtils.hs]: You should make a longer note here too.

File: compiler/deSugar/DsUtils.hs [hunk 7]
```diff
@@ -116,17 +116,18 @@ selectSimpleMatchVarL pat = selectMatchVar (unLoc pat)
...
+selectMatchVar w (BangPat _ pat) = selectMatchVar w (unLoc pat)
+selectMatchVar w (LazyPat _ pat) = selectMatchVar w (unLoc pat)
+selectMatchVar w (ParPat _ pat)  = selectMatchVar w (unLoc pat)
+selectMatchVar w (VarPat _ var)  = return (localiseId (unLoc var))
                                   -- Note [Localise pattern binders]
-selectMatchVar (AsPat _ var _) = return (unLoc var)
-selectMatchVar other_pat     = newSysLocalDsNoLP Omega (hsPatType other_pat) -- TODO: arnaud: I'm pretty sure I actually need to take the multiplicity of the pattern as an argument, though not knowing what the variable is used for, I don't know how I would use it quite yet
+selectMatchVar w (AsPat _ var _) = return (unLoc var)
+selectMatchVar w other_pat     =
+    newSysLocalDsNoLP w (hsPatType other_pat) -- TODO: arnaud: I'm pretty sure I actually need to take the multiplicity of the pattern as an argument, though not knowing what the variable is used for, I don't know how I would use it quite yet
```
Comment (aspiwack) [inline compiler/deSugar/DsUtils.hs]: You can remove my note here.

File: compiler/deSugar/Match.hs [hunk 8]
```diff
@@ -161,18 +161,19 @@ type MatchId = Id   -- See Note [Match Ids]
 match :: [MatchId]        -- Variables rep\'ing the exprs we\'re matching with
                           -- See Note [Match Ids]
       -> Type             -- Type of the case expression
+      -> Rig              -- Weight of the case expression
```
Comment (aspiwack) [inline compiler/deSugar/Match.hs]: Lets say « Multiplicity » instead of « Weight » here. (we will change the names of the modules and types later, but let's get the comments aligned with the paper and proposal).

File: compiler/deSugar/Match.hs [hunk 9]
```diff
@@ -228,70 +230,73 @@ match vars@(v:_) ty eqns    -- Eqns *can* be empty
...
-matchCoercion :: HasCallStack => [MatchId] -> Type -> [EquationInfo] -> DsM MatchResult
+matchCoercion :: HasCallStack => [MatchId] -> Type -> Rig -> [EquationInfo] -> DsM MatchResult
 -- Apply the coercion to the match variable and then match that
-matchCoercion (var:vars) ty (eqns@(eqn1:_))
+matchCoercion (var:vars) ty weight (eqns@(eqn1:_))
   = do  { let CoPat _ co pat _ = firstPat eqn1
         ; let pat_ty' = hsPatType pat
-        ; var' <- newUniqueId var (idWeight var) pat_ty'
-        ; match_result <- match (var':vars) ty $
+        ; var' <- newUniqueId var Omega pat_ty'
```
Comment (aspiwack) [inline compiler/deSugar/Match.hs]: Is `Omega` correct here? If so, I suggest making a comment explaining why.

File: compiler/hsSyn/HsBinds.hs [hunk 10]
```diff
@@ -679,7 +679,7 @@ pprDeclList :: [SDoc] -> SDoc   -- Braces with a space
 pprDeclList ds = pprDeeperList vcat ds
 
 ------------
-emptyLocalBinds :: HsLocalBindsLR (GhcPass a) (GhcPass b)
+emptyLocalBinds :: (XEmptyLocalBinds a b ~ PlaceHolder) => HsLocalBindsLR a b
```
Comment (aspiwack) [inline compiler/hsSyn/HsBinds.hs]: I'm confused about this one. It looks completely unrelated. What happened here?

File: compiler/hsSyn/HsUtils.hs [hunk 11]
```diff
@@ -143,9 +144,10 @@ just attach noSrcSpan to everything.
 mkHsPar :: LHsExpr (GhcPass id) -> LHsExpr (GhcPass id)
 mkHsPar e = L (getLoc e) (HsPar noExt e)
 
-mkSimpleMatch :: HsMatchContext (NameOrRdrName (IdP (GhcPass p)))
-              -> [LPat (GhcPass p)] -> Located (body (GhcPass p))
-              -> LMatch (GhcPass p) (Located (body (GhcPass p)))
+mkSimpleMatch :: (XEmptyLocalBinds p p ~ PlaceHolder)
+               => HsMatchContext (NameOrRdrName (IdP p))
+              -> [LPat p] -> Located (body p)
+              -> LMatch p (Located (body p))
```
Comment (aspiwack) [inline compiler/hsSyn/HsUtils.hs]: Though the changes in this file look related.  I'd love to understand.

File: compiler/typecheck/TcArrows.hs [hunk 12]
```diff
@@ -259,7 +259,8 @@ tc_cmd env
                                          , m_grhss = grhss' })
               arg_tys = map hsLPatType pats'
               cmd' = HsCmdLam x (MG { mg_alts = L l [match'], mg_arg_tys = map unrestricted arg_tys
-                                  , mg_res_ty = res_ty, mg_origin = origin })
+                                  , mg_res_ty = res_ty, mg_origin = origin, mg_weight = Omega  })
+                                    -- MattP: Check
```
Comment (aspiwack) [inline compiler/typecheck/TcArrows.hs]: Longer note, maybe?

File: compiler/deSugar/Match.hs (hunk 9 above)
Comment (mpickering) [reply compiler/deSugar/Match.hs]: No! I think that is a mistake.

File: compiler/hsSyn/HsBinds.hs (hunk 10 above)
Comment (mpickering) [reply compiler/hsSyn/HsBinds.hs]: Ah, these changes snuck in when I modified the wrong data type originally. I'll remove them from this patch but they are to do with fixing https://ghc.haskell.org/trac/ghc/ticket/15084

**Comment by aspiwack** [comment]
By the way, I didn't realise, but why are we merging into `wip/linear-types`, rather than `linear-types`?

**Comment by mpickering** [comment]
I thought that we wanted to keep `linear-types` green? The linearity checks are only enabled on the `wip/linear-types` branch right now.

**Comment by aspiwack** [comment]
I see.

**Comment by aspiwack** [comment]
An alternative is to enable the linearity check in linear-types and deactivate the tests that it breaks, and reactivate them when we fix them.
